File: src/web/routes/websocket.py
"""
WebSocket endpoint for real-time streaming transcription.
Supports long-duration recordings with server-side chunk persistence.
"""

import json
import base64
import asyncio
from datetime import datetime
from typing import Optional
from fastapi import APIRouter, WebSocket, WebSocketDisconnect

from ..services.streaming_transcriber import StreamingTranscriber
from ..services.recording_session import SessionManager, RecordingSession
from ..services.history_manager import HistoryManager

router = APIRouter()

# Session manager for long recordings
session_manager = SessionManager()


async def finalize_with_keepalive(websocket: WebSocket, transcriber: StreamingTranscriber):
    """
    Run finalize() while sending periodic keep-alive messages.
    This prevents the WebSocket from timing out during long transcription.
    """
    # Create a task for finalization
    finalize_task = asyncio.create_task(transcriber.finalize())

    # Send keep-alive while waiting
    while not finalize_task.done():
        try:
            await asyncio.wait_for(asyncio.shield(finalize_task), timeout=1.0)
        except asyncio.TimeoutError:
            # Send keep-alive ping
            try:
                await websocket.send_json({"type": "ping"})
            except Exception:
                pass  # Connection might be closing

    return finalize_task.result()


@router.websocket("/transcribe")
async def websocket_transcribe(websocket: WebSocket):
    """
    WebSocket endpoint for real-time audio transcription.

    Protocol:
    1. Client connects
    2. Client sends 'start' message with config: {"type": "start", "model": "base", "language": null}
    3. Server responds with 'ready' message (includes session_id for long recordings)
    4. Client sends 'audio' messages: {"type": "audio", "data": "<base64-encoded-pcm>"}
    5. Server sends 'transcript' messages as transcription becomes available
    6. Client can send 'flush' messages to persist audio on server: {"type": "flush", "data": "<base64>"}
    7. Client sends 'stop' message to finalize
    8. Server sends 'complete' message with final transcript
    """
    await websocket.accept()

    transcriber: Optional[StreamingTranscriber] = None
    session: Optional[RecordingSession] = None
    current_model: str = "base"
    current_language: Optional[str] = None

    try:
        while True:
            # Receive message
            try:
                data = await websocket.receive_text()
                message = json.loads(data)
            except json.JSONDecodeError:
                await websocket.send_json({
                    "type": "error",
                    "error": "Invalid JSON message",
                })
                continue

            msg_type = message.get("type")

            if msg_type == "start":
                # Initialize transcriber with config
                current_model = message.get("model", "base")
                current_language = message.get("language")
                sample_rate = message.get("sample_rate", 16000)
                chunk_duration = message.get("chunk_duration", 5.0)
                enable_persistence = message.get("enable_persistence", True)

                transcriber = StreamingTranscriber(
                    model_size=current_model,
                    language=current_language,
                    sample_rate=sample_rate,
                    chunk_duration=chunk_duration,
                )

                # Create session for long recording support
                if enable_persistence:
                    session = session_manager.create_session(sample_rate=sample_rate)
                    session_id = session.session_id
                else:
                    session_id = None

                await websocket.send_json({
                    "type": "ready",
                    "message": f"Transcriber initialized with model '{current_model}'",
                    "session_id": session_id,
                })

            elif msg_type == "audio" and transcriber:
                # Decode base64 audio data
                audio_base64 = message.get("data", "")
                try:
                    audio_bytes = base64.b64decode(audio_base64)
                except Exception:
                    await websocket.send_json({
                        "type": "error",
                        "error": "Invalid base64 audio data",
                    })
                    continue

                # Process audio chunk
                result = await transcriber.process_audio_chunk(audio_bytes)

                if result and result.get("text"):
                    await websocket.send_json({
                        "type": "transcript",
                        "text": result["text"],
                        "is_final": result.get("is_final", False),
                        "timestamp": result.get("timestamp", 0),
                        "confidence": result.get("confidence", 0),
                    })

            elif msg_type == "flush" and session:
                # Persist audio chunk to server-side storage
                audio_base64 = message.get("data", "")
                try:
                    audio_bytes = base64.b64decode(audio_base64)
                    chunk = session.add_chunk(audio_bytes)

                    await websocket.send_json({
                        "type": "flush_ack",
                        "success": True,
                        "chunk_duration": chunk.duration,
                        "total_duration": session.total_duration,
                        "chunk_count": len(session.chunks),
                    })
                except Exception as e:
                    await websocket.send_json({
                        "type": "flush_ack",
                        "success": False,
                        "error": str(e),
                    })

            elif msg_type == "session_stats" and session:
                # Return session statistics
                stats = session.get_stats()
                await websocket.send_json({
                    "type": "session_stats",
                    **stats,
                })

            elif msg_type == "stop":
                # Finalize transcription
                if transcriber:
                    # Send immediate acknowledgment
                    await websocket.send_json({
                        "type": "status",
                        "status": "processing",
                        "message": "Finalizing transcription, please wait...",
                    })

                    # If we have persisted chunks, add them to transcriber for final processing
                    if session and len(session.chunks) > 0:
                        # Save session audio to temp file for complete re-transcription
                        await websocket.send_json({
                            "type": "status",
                            "status": "processing",
                            "message": f"Processing {session.total_duration:.0f}s of recorded audio...",
                        })

                    # Process final transcription with keep-alive pings
                    final_result = await finalize_with_keepalive(websocket, transcriber)

                    await websocket.send_json({
                        "type": "complete",
                        "text": final_result.get("text", ""),
                        "is_final": True,
                        "segments": final_result.get("segments", []),
                        "session_duration": session.total_duration if session else 0,
                    })

                    # Save to history
                    try:
                        final_text = final_result.get("text", "")
                        if final_text.strip():
                            history_manager = HistoryManager()
                            # Create result dict matching save_transcription format
                            result_for_history = {
                                "text": final_text,
                                "language": current_language,
                                "duration": session.total_duration if session else 0,
                                "confidence": 0.9,
                                "segments": final_result.get("segments", []),
                                "metadata": {
                                    "transcription": {
                                        "model": current_model,
                                    }
                                }
                            }
                            history_manager.save_transcription(
                                result=result_for_history,
                                filename=f"Recording_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
                            )
                    except Exception as hist_err:
                        print(f"Warning: Failed to save recording to history: {hist_err}")

                    # Cleanup
                    transcriber.cleanup()
                    transcriber = None
                else:
                    # No transcriber, send empty result
                    await websocket.send_json({
                        "type": "complete",
                        "text": "",
                        "is_final": True,
                        "segments": [],
                    })

                break

            elif msg_type == "ping":
                # Keep-alive ping
                await websocket.send_json({"type": "pong"})

            else:
                await websocket.send_json({
                    "type": "error",
                    "error": f"Unknown message type: {msg_type}",
                })

    except WebSocketDisconnect:
        pass
    except Exception as e:
        try:
            await websocket.send_json({
                "type": "error",
                "error": str(e),
            })
        except Exception:
            pass
    finally:
        if transcriber:
            transcriber.cleanup()
        if session:
            session_manager.remove_session(session.session_id)

File: src/web/static/js/websocket.js
/**
 * WebSocket client for real-time transcription.
 */

class TranscriptionWebSocket {
    constructor() {
        this.socket = null;
        this.isConnected = false;
        this.reconnectAttempts = 0;
        this.maxReconnectAttempts = 3;
        this.fullTranscript = '';

        // Callbacks for long recording support
        this.onReady = null;
        this.onFlushAck = null;
        this.onSessionStats = null;
    }

    /**
     * Connect to the WebSocket server.
     */
    async connect() {
        return new Promise((resolve, reject) => {
            const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
            const url = `${protocol}//${window.location.host}/ws/transcribe`;

            this.socket = new WebSocket(url);

            this.socket.onopen = () => {
                console.log('WebSocket connected');
                this.isConnected = true;
                this.reconnectAttempts = 0;
                resolve();
            };

            this.socket.onclose = (event) => {
                console.log('WebSocket closed:', event.code, event.reason);
                this.isConnected = false;

                // If we have accumulated transcript, show it even if connection closed unexpectedly
                if (this.fullTranscript.trim() && !event.wasClean) {
                    console.log('Connection closed unexpectedly, showing accumulated transcript');
                    const result = {
                        text: this.fullTranscript.trim(),
                        language: 'auto',
                        confidence: 0.8,
                        processing_time: 0,
                        segments: [],
                    };

                    if (typeof hideLiveTranscript === 'function') {
                        hideLiveTranscript();
                    }
                    showResults(result);
                    if (typeof showRecordingActions === 'function') {
                        showRecordingActions();
                    }
                }
            };

            this.socket.onerror = (error) => {
                console.error('WebSocket error:', error);
                reject(error);
            };

            this.socket.onmessage = (event) => {
                this.handleMessage(event.data);
            };
        });
    }

    /**
     * Send a message to the server.
     */
    send(data) {
        if (this.socket && this.isConnected) {
            this.socket.send(JSON.stringify(data));
        }
    }

    /**
     * Handle incoming messages.
     */
    handleMessage(data) {
        try {
            const message = JSON.parse(data);

            switch (message.type) {
                case 'ready':
                    console.log('Transcriber ready:', message.message);
                    if (message.session_id) {
                        console.log('Session ID:', message.session_id);
                    }
                    // Call callback if set
                    if (this.onReady) {
                        this.onReady(message);
                    }
                    break;

                case 'transcript':
                    this.handleTranscript(message);
                    break;

                case 'complete':
                    this.handleComplete(message);
                    break;

                case 'status':
                    console.log('Status:', message.status, message.message);
                    // Show processing status to user
                    if (message.status === 'processing') {
                        updateLiveTranscript(message.message || 'Processing final transcription...');
                    }
                    break;

                case 'flush_ack':
                    // Acknowledgment that server received and stored flushed audio
                    if (message.success) {
                        console.log(`Flush acknowledged: ${message.chunk_count} chunks, ${message.total_duration?.toFixed(1)}s total`);
                    } else {
                        console.error('Flush failed:', message.error);
                    }
                    if (this.onFlushAck) {
                        this.onFlushAck(message);
                    }
                    break;

                case 'session_stats':
                    // Session statistics from server
                    console.log('Session stats:', message);
                    if (this.onSessionStats) {
                        this.onSessionStats(message);
                    }
                    break;

                case 'error':
                    console.error('Server error:', message.error);
                    showError(message.error);
                    break;

                case 'pong':
                case 'ping':
                    // Keep-alive response
                    break;

                default:
                    console.log('Unknown message type:', message.type);
            }
        } catch (error) {
            console.error('Failed to parse message:', error);
        }
    }

    /**
     * Handle transcript update.
     */
    handleTranscript(message) {
        if (message.text) {
            // Append to full transcript
            this.fullTranscript += ' ' + message.text;

            // Update live display
            updateLiveTranscript(this.fullTranscript.trim());
        }
    }

    /**
     * Handle transcription complete.
     */
    handleComplete(message) {
        console.log('Transcription complete:', message);

        // Log session duration for long recordings
        if (message.session_duration) {
            const minutes = Math.floor(message.session_duration / 60);
            const seconds = Math.floor(message.session_duration % 60);
            console.log(`Recording session duration: ${minutes}m ${seconds}s`);
        }

        // Show final result
        const result = {
            text: message.text || this.fullTranscript.trim(),
            language: 'auto',
            confidence: 0.9,
            processing_time: 0,
            segments: message.segments || [],
            session_duration: message.session_duration || 0,
        };

        // Hide live transcript, show results
        if (typeof hideLiveTranscript === 'function') {
            hideLiveTranscript();
        }

        // Show results and recording actions
        showResults(result);

        // Show download buttons for recording mode
        if (typeof showRecordingActions === 'function') {
            showRecordingActions();
        }

        // Now close the WebSocket connection
        this.close();
    }

    /**
     * Close the WebSocket connection.
     */
    close() {
        if (this.socket) {
            this.socket.close();
            this.socket = null;
        }
        this.isConnected = false;
        this.fullTranscript = '';
    }

    /**
     * Send keep-alive ping.
     */
    ping() {
        this.send({ type: 'ping' });
    }
}
